typescript_typeorm_graphql_project/src/controllers/auth.ts

```diff
@@ -0,0 +1,104 @@
+import { validationResult } from "express-validator";
+import { hash, compare } from "bcryptjs";
+import { sign } from "jsonwebtoken";
+import * as dotenv from "dotenv";
+import { Request, Response, NextFunction } from "express";
+import { User } from "../entities/User";
+dotenv.config();
+
+const signUp = async (
+  req: Request | any,
+  res: Response,
+  next: NextFunction
+) => {
+  try {
+    const errors = validationResult(req);
+    console.log(req.body);
+
+    if (!errors.isEmpty()) {
+      res.statusCode = 422;
+      const error = new Error("Validation Failed.");
+      throw error;
+    }
+
+    const email: string = req.body.email;
+    const name: string = req.body.name;
+    const password: string = req.body.password;
+    const role: string = req.body.role;
+
+    const hashPassword = await hash(password, 12);
+
+    console.log(hashPassword);
+
+    const user = new User();
+
+    user.email = email;
+    user.name = name;
+    user.role = role;
+    user.password = hashPassword;
+
+    const result = await user.save();
+
+    res
+      .status(200)
+      .json({ message: "User created!", userId: result.id, userRole: role });
+    return result;
+  } catch (err: any) {
+    if (!err.statusCode) {
+      err.statusCode = 500;
+    }
+    next(err);
+  }
+};
+
+const login = async (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const email = req.body.email;
+    const password = req.body.password;
+
+    const user = await User.findOne({
+      where: {
+        email: email,
+      },
+    });
+
+    if (!user) {
+      res.statusCode = 401;
+      const error = new Error("User with this email could not found.");
+      throw error;
+    }
+
+    const isValid: boolean = await compare(password, user.password);
+
+    if (!isValid) {
+      res.statusCode = 401;
```

Comment:
If user is not exists, then it should 404 error statusCode, not 401
401 is for not authorized

Reply:
404 is for bad request. 401 status code indicates that the request has not been applied because it lacks valid authentication credentials for the target resource.